### DISCUSSION TITLE
Is there a replacement for FileCodeModel and CodeElements?

### DISCUSSION THREAD

**Discussion by trihugger** [VsixCommunity/Community.VisualStudio.Toolkit#308] [Q&A] [answered]
Hello,

I'm used to using EnvDTE.FileCodeModel to extract code elements (EnvDTE.CodeElements) from a code file.

I looked at DocumentView, which exposes TextBuffer, CurrentSnapshot, and ITextSnapshotLine, but i'm looking for a better way to break down the elements from each code line like CodeElements provides.

Thank you,

Jeff

**Comment by reduckted** [comment]
Depending on your use case, you might be able to use the Roslyn workspace.

https://stackoverflow.com/a/28199656/4397397

That will only work for C# and VB.NET files, and is dramatically different to the file code model of Visual Studio, so it may be completely unsuitable for you, but I thought I'd mention it anyway. 😄

**Comment by LesterSmith** [comment, marked as answer]
It appears that you can still get to the FileCodeModel from the DTE object like the following code, even in **VS2022**
    var codeElts = dte.ActiveDocument.ProjectItem.FileCodeModel.CodeElements;

I am able to get the DTE object in VS2022 from the InitializeAsync method of a VSPackage without adding an Editor Adornment item (which causes unwanted effects if you are not trying to change the appearance of the editor).  I created a VSIX project with a command and then added a Package.  To get the DTE object, I added the following code to the InitializeAsync method of the Package.

            // switch to main thread to get the DTE object
            await this.JoinableTaskFactory.SwitchToMainThreadAsync();

            // use this package's GetServiceAsync and save DTE in a
            // static Globals Class for later use
            // you cannot do anything with it here except save it
            // you can later use it in the command execute method
            DTE dte = await this.GetServiceAsync(typeof(DTE)) as DTE;
            
            // save the DTE in a Global Static class; it is immutable
            Globals.dTE = dte;

            // this works too, using the static GetGlobalService (only works in the Package initialization)
            DTE dte1 = (DTE)GetGlobalService(typeof(DTE));

I created a Globals class to save the DTE object in because you will need it later in the command Execute method.

 using EnvDTE;
 internal class Globals
 {
  public static DTE dTE { get; set; }

  public static DTE dTE2 { get; set; }
 }

Finally, in the ExecuteAsync method of the Command Class I inserted the following code to test the FileCodeModel:

            if (Globals.dTE != null)
            {
                DTE dte = Globals.dTE;
                if (dte.ActiveDocument.Name.ToLower().EndsWith(".cs") || dte.ActiveDocument.Name.ToLower().EndsWith(".vb"))
                {
                    var sb = new System.Text.StringBuilder();
                    var codeElts = dte.ActiveDocument.ProjectItem.FileCodeModel.CodeElements;
                    foreach (CodeElement codeElement in codeElts)
                    {
                        sb.Append($"CodeElement {codeElement.Name} is Kind {codeElement.Kind}" + Environment.NewLine);
                    }
                    System.Diagnostics.Debug.WriteLine(sb.ToString());
                }
            }
